File: Ejercicio 10/Ejercicio 10.cs
using System;

public class HeapSort
{
    public static void HeapSortArray(int[] arr)
    {
        int n = arr.Length;
        
        for (int i = n / 2 - 1; i >= 0; i--)
        {
            Heapify(arr, n, i);
        }
        
        for (int i = n - 1; i > 0; i--)
        {
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            
            Heapify(arr, i, 0);
        }
    }

    static void Heapify(int[] arr, int n, int i)
    {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        
        if (left < n && arr[left] > arr[largest])
        {
            largest = left;
        }
        
        if (right < n && arr[right] > arr[largest])
        {
            largest = right;
        }
        
        if (largest != i)
        {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;
            
            Heapify(arr, n, largest);
        }
    }
    
    static void PrintArray(int[] arr)
    {
        for (int i = 0; i < arr.Length; i++)
        {
            Console.Write(arr[i] + " ");
        }
        Console.WriteLine();
    }
    
    static int[] GenerarArrayAleatorio(int tamaño, int min, int max)
    {
        int[] arr = new int[tamaño];
        Random random = new Random();
        
        for (int i = 0; i < tamaño; i++)
        {
            arr[i] = random.Next(min, max + 1);
        }
        return arr;
    }
    
    public static void Main(string[] args)
    {
        int[] arr = GenerarArrayAleatorio(15, 1, 100);
        
        Console.WriteLine("Array original (15 valores aleatorios):");
        PrintArray(arr);
        
        HeapSortArray(arr);
        
        Console.WriteLine("\nArray ordenado con Heap Sort:");
        PrintArray(arr);
    }
}

File: Ejercicio 10/Ejercicio 10.cpp
#include <iostream>
#include <vector>
#include <random>
#include <ctime>

using namespace std;

void heapify(vector<int>& arr, int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    
    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }
    
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }
    
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();
    
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }
    
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

void printArray(const vector<int>& arr) {
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

vector<int> generarArrayAleatorio(int tamaño, int min, int max) {
    vector<int> arr(tamaño);
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> distrib(min, max);
    
    for (int i = 0; i < tamaño; i++) {
        arr[i] = distrib(gen);
    }
    return arr;
}

int main() {
    vector<int> arr = generarArrayAleatorio(15, 1, 100);
    
    cout << "Array original (15 valores aleatorios):" << endl;
    printArray(arr);
    
    heapSort(arr);
    
    cout << "\nArray ordenado con Heap Sort:" << endl;
    printArray(arr);
    
    return 0;
}